pkg/clients/rds/rds.go

```diff
@@ -452,7 +458,50 @@ func IsUpToDate(p v1beta1.RDSInstanceParameters, db rds.DBInstance) (bool, error
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "Tags"),
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "SkipFinalSnapshotBeforeDeletion"),
 		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "FinalDBSnapshotIdentifier"),
-	), nil
+		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "ApplyModificationsImmediately"),
+		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "AllowMajorVersionUpgrade"),
+		cmpopts.IgnoreFields(v1beta1.RDSInstanceParameters{}, "MasterPasswordSecretRef"),
+	) && !pwdChanged, nil
+}
+
+// GetPassword fetches the referenced input password for an RDSInstance CRD and determines whether it has changed or not
+func GetPassword(ctx context.Context, kube client.Client, r *v1beta1.RDSInstance) (string, bool, error) {
+	if kube == nil {
+		return "", false, nil
+	}
+	if r.Spec.ForProvider.MasterPasswordSecretRef == nil && r.Spec.WriteConnectionSecretToReference == nil {
```

Comment:
```suggestion
	if r.Spec.ForProvider.MasterPasswordSecretRef == nil || r.Spec.WriteConnectionSecretToReference == nil {
```

In order for this function to make its job, both fields have to be set. So, if any of them isn't there, then we should return early. As written here, that's not an error case, it's just saying that we don't need the utility of `GetPassword` for those cases.

Reply:
After you make this change, in the rest of the function you can be sure that they are not `nil` so you can get rid of two `if` conditions below.

Reply:
Neither `masterPasswordSecretRef` (input) nor `writeConnectionSecretToRef` (output) are marked as required, so I figure we should try and account for situations where either are set. I figure the best way to dissect this would be to walk through the contexts in which this function may be called.

-----

### Create

When creating an instance, we only really check the input. If input is set, we'll use that value for the password we provide to AWS. If the input field is not set, or the value in the referenced secret is set to `""` we'll autogenerate a password. I've created a small diagram for [it here](https://i.imgur.com/Qn7BJTL.png).

If we change the `GetPassword` function to return if either input or output references are nil, we wouldn't report back what the input password is so I'd revert the change in `Create` and have it pull the input secret directly.

### Update

I created [another diagram](https://i.imgur.com/kEfYaRR.png) to help convey my thought process. As you can see, we don't really do anything unless both fields are set, so I can see how making this change would clean up the code a bit.

### Observe

I created [a diagram](https://i.imgur.com/pPM4rCz.png) to visually represent this flow as well. Again we only really do work if both fields are set.

-----

If we have `Create` pull the password directly as it was doing before, I can 100% support cleaning up the code a bit. I'd also feel like a rename from `GetPassword` to `CheckPassword` as suggested in your original round of reviews would completely describe what it is doing. Otherwise, I feel we should at least continue supporting fields being partially set, to account for the `Create` case.

Reply:
Thanks for all the details! I think I understand what you're describing. Seems like we have only one case that's tricky; input is given but output is not given during `Create`. I couldn't really think of many scenarios where you give an input secret (meaning you use pwd auth) but not use the output. However, for the implementation to be _correct_, i.e. user supplied a password and that should be enough for us to use it, I agree with you that this should be taken into account.

So, in all cases, if input is empty we should return early but output being empty is not of a concern so that we can take `Create` into account. So, I'd suggest:
```suggestion
	if r.Spec.ForProvider.MasterPasswordSecretRef == nil {
```

and assume that `r.Spec.ForProvider.MasterPasswordSecretRef` is not `nil` in the rest of the function. The check for `if r.Spec.WriteConnectionSecretToReference != nil {` in the later parts of the function seems necessary.

Reply:
Pushed this up